sympy/crypto/crypto.py

```diff
@@ -2037,3 +2038,200 @@ def dh_shared_key(key, b):
+
+
+def _legendre(a, p):
+    """
+    Returns the legendre symbol of a and p
+    assuming that p is a prime
+
+    i.e. 1 if a is a quadratic residue mod p
+        -1 if a is not a quadratic residue mod p
+         0 if a is divisible by p
+
+    Parameters
+    ==========
+
+    a : int the number to test
+    p : the prime to test a against
+
+    Returns
+    =======
+
+    legendre symbol (a / p) (int)
+
+    """
+    sig = pow(a%p, (p - 1)//2) % p
+    if sig == 1:
+        return 1
+    elif sig == 0:
+        return 0
+    else:
+        return -1
+
+
+def _random_coprime_stream(n, seed=None):
+    randrange = _randrange(seed)
+    while True:
+        y = randrange(n)
+        if gcd(y, n) == 1:
+            yield y
+
+
+def gm_private_key(p, q, a=None):
+    """
+    Check if p and q can be used as private keys for
+    the Goldwasser-Micali encryption. The method works
+    roughly as follows.
+
+    Pick two large primes p ands q. Call their product N.
+    Given a message as an integer i, write i in its
+    bit representation b_0,...,b_n. For each k,
+
+     if b_k = 0:
+        let a_k be a random square
+        (quadratic residue) modulo p * q
+        such that jacobi_symbol(a, p * q) = 1
+     if b_k = 1:
+        let a_k be a random non-square
+        (non-quadratic residue) modulo p * q
+        such that jacobi_symbol(a, p * q) = 1
+
+    return [a_1, a_2,...]
+
+    b_k can be recovered by checking whether or not
+    a_k is a residue. And from the b_k's, the message
+    can be reconstructed.
+
+    The idea is that, while jacobi_symbol(a, p * q)
+    can be easily computed (and when it is equal to -1 will
+    tell you that a is not a square mod p * q), quadratic
+    residuosity modulo a composite number is hard to compute
+    without knowing its factorization.
+
+    Moreover, approximately half the numbers coprime to p * q have
+    jacobi_symbol equal to 1. And among those, approximately half
+    are residues and approximately half are not. This maximizes the
+    entropy of the code.
+
+    Parameters
+    ==========
+
+    p, q, a : initialization variables
+
+    Returns
+    =======
+
+    p, q | false (int)
+        if inputs cannot be used to initalize public key
+        then refurn false. Otherwise, return p and q
+
+    """
+    if p == q:
+        return False
+    elif not isprime(p) or not isprime(q):
+        return False
+    elif p == 2 or q == 2:
+        return False
+    return p, q
+
+
+def gm_public_key(p, q, a=None, seed=None):
+    """
+    Compute public keys for p and q.
+    Note that in Goldwasser-Micali Encrpytion,
+    public keys are randomly selected.
+
+    Parameters
+    ==========
+
+    p, q, a : (int) initialization variables
+
+    Returns
+    =======
+
+    a | false (int)
+        if a is not None otherwise some random integer
+        coprime to p and q. returns false
+        if inputs cannot be used to initalize public key
+
+    N (int) the product of p and q
```

Comment:
Make it clearer here that it returns a tuple.

Reply:
Is there anything else that should be done?